core/src/main/java/com/google/errorprone/internal/NonDelegatingClassLoader.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.errorprone.internal;
+
+import com.google.common.collect.ImmutableSet;
+
+import java.net.URLClassLoader;
+import java.util.Set;
+
+/**
+ * A non-delegating {@link java.net.URLClassLoader} that searches it's own resource path _before_
+ * the runtime classpath, reversing the usual classloader delegation model. This makes it possible
+ * to override runtime classes.
+ *
+ * <p>The classloader is also given the classloader from the environment it is created in,
+ * and a whitelist of classes that it should load from that environment. This list must
+ * include anything that crosses classloader boundaries (for example, the input or return types
+ * of the callback).
+ *
+ * @author cushon@google.com
+ */
+public class NonDelegatingClassLoader extends URLClassLoader {
+  private final ClassLoader original;
+  private final ImmutableSet<String> whiteList;
+
+  public static NonDelegatingClassLoader create(Set<String> whiteList) {
+    ClassLoader contextLoader = Thread.currentThread().getContextClassLoader();
+    if (!(contextLoader instanceof URLClassLoader)) {
```

Comment:
This assumption is unfortunate, but it holds for all known plugins and I don't know any reasonable alternatives.